src/host/directio.cpp

```diff
@@ -237,35 +138,92 @@ void EventsToUnicode(_Inout_ std::deque<std::unique_ptr<IInputEvent>>& inEvents,
                                                           const std::span<const INPUT_RECORD> buffer,
                                                           size_t& written,
                                                           const bool append) noexcept
+try
 {
     written = 0;
 
+    if (buffer.empty())
+    {
+        return S_OK;
+    }
+
     LockConsole();
     auto Unlock = wil::scope_exit([&] { UnlockConsole(); });
 
-    try
+    const auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+    til::small_vector<INPUT_RECORD, 16> events;
+
+    // Check out the loop below. When a previous call ended on a leading DBCS we store it for
+    // the next call to WriteConsoleInputAImpl to join it with the now available trailing DBCS.
+    if (context.IsWritePartialByteSequenceAvailable())
     {
-        auto events = IInputEvent::Create(buffer);
+        auto lead = context.FetchWritePartialByteSequence(false)->ToInputRecord();
+        const auto& trail = buffer.front();
+
+        if (trail.EventType == KEY_EVENT)
+        {
+            const char narrow[2]{
+                lead.Event.KeyEvent.uChar.AsciiChar,
+                trail.Event.KeyEvent.uChar.AsciiChar,
+            };
+            wchar_t wide[2];
+            const auto length = MultiByteToWideChar(gci.CP, 0, &narrow[0], 2, &wide[0], 2);
+
+            for (int i = 0; i < length; i++)
+            {
+                lead.Event.KeyEvent.uChar.UnicodeChar = wide[i];
+                events.push_back(lead);
+            }
+        }

```

Comment:
hey, we never remove `buffer[0]` from consideration. Doesn't that mean we process it two times?

Reply:
(here, and in the loop on R180?)

Reply:
If I turned out to be correct... please add a test to make sure that it worked as expected ;)